File: core/src/com/alexpi/awesometanks/entities/blocks/Box.java
package com.alexpi.awesometanks.entities.blocks;

import com.alexpi.awesometanks.entities.DamageListener;
import com.alexpi.awesometanks.entities.items.FreezingBall;
import com.alexpi.awesometanks.entities.items.GoldNugget;
import com.alexpi.awesometanks.entities.items.HealthPack;
import com.alexpi.awesometanks.entities.tanks.EnemyTank;
import com.alexpi.awesometanks.utils.Utils;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.PolygonShape;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.scenes.scene2d.Group;

/**
 * Created by Alex on 19/01/2016.
 */
public class Box extends Block {
    private AssetManager manager;
    private Vector2 targetPosition;
    private int maxType;
    public Box(DamageListener listener, AssetManager manager, World world, Vector2 targetPosition, Vector2 pos, int level){
        super(manager,"sprites/box.png", world,new PolygonShape(),50, pos, .8f, true, listener);
        this.manager = manager;
        this.targetPosition = targetPosition;
        maxType = Spawner.getMaxType(level);
    }

    public void drop(){
        switch (Utils.getRandomInt(4)){
            case 0:
                int num1 = Utils.getRandomInt(5,16);
                for(int i =0; i <num1;i++)
                    getParent().addActor(new GoldNugget(manager,body.getWorld(),body.getPosition()));
                break;
            case 1:
                getParent().addActor(new FreezingBall(manager, body.getWorld(), body.getPosition()));break;
            case 2:
                getParent().addActor(new HealthPack(manager,body.getWorld(),body.getPosition()));break;
            case 3:
                getParent().addActor(new EnemyTank(manager, body.getWorld(), body.getPosition(),targetPosition, EnemyTank.Tier.MINI, Utils.getRandomInt(maxType +1), getDamageListener()));break;
        }
    }

    @Override public void detach() {
        drop();
        super.detach();

    }
}

File: core/src/com/alexpi/awesometanks/entities/blocks/Box.kt
package com.alexpi.awesometanks.entities.blocks

import com.alexpi.awesometanks.entities.components.body.BodyShape
import com.alexpi.awesometanks.entities.components.body.FixtureFilter
import com.alexpi.awesometanks.entities.items.FreezingBall
import com.alexpi.awesometanks.entities.items.GoldNugget
import com.alexpi.awesometanks.entities.items.HealthPack
import com.alexpi.awesometanks.entities.tanks.EnemyTank
import com.alexpi.awesometanks.screens.LEVEL_COUNT
import com.alexpi.awesometanks.screens.game.stage.GameContext
import com.alexpi.awesometanks.utils.RandomUtils
import com.alexpi.awesometanks.weapons.Weapon
import com.badlogic.gdx.math.Vector2

/**
 * Created by Alex on 19/01/2016.
 */
class Box(private val gameContext: GameContext, level: Int, pos: Vector2) :
    HealthBlock(gameContext, "sprites/box.png", BodyShape.Box(.8f, .8f), pos,50f,  true, false, FixtureFilter.BLOCK) {
    private var generatedTypes: List<Weapon.Type> = getEnemyTypes(level)
    private val nuggetValue: Int = getNuggetValue(level)
    private fun dropLoot() {
        when (RandomUtils.getRandomInt(4)) {
            0 -> {
                val num1 = RandomUtils.getRandomInt(10, 16)
                var i = 0
                while (i < num1) {
                    parent.addActor(
                        GoldNugget(
                            bodyComponent.body.position,
                            RandomUtils.getRandomInt(nuggetValue - 5, nuggetValue + 5)
                        )
                    )
                    i++
                }
            }

            1 -> parent.addActor(FreezingBall(bodyComponent.body.position))
            2 -> parent.addActor(HealthPack(bodyComponent.body.position))
            3 -> parent.addActor(
                EnemyTank(
                    gameContext,
                    bodyComponent.body.position,
                    EnemyTank.Tier.MINI,
                    generatedTypes.random()
                )
            )
        }
    }

    override fun remove(): Boolean {
        dropLoot()
        return super.remove()
    }

    companion object {
        private fun getNuggetValue(level: Int): Int {
            return (20 + level.toFloat() / (LEVEL_COUNT - 1) * 50).toInt()
        }

        private fun getMaxType(level: Int): Int {
            return if (level <= 7) Weapon.Type.RICOCHET.ordinal
            else if (level <= 15) Weapon.Type.ROCKETS.ordinal
            else if (level <= 22) Weapon.Type.LASERGUN.ordinal
            else Weapon.Type.RAILGUN.ordinal
        }

        private fun getEnemyTypes(level: Int): List<Weapon.Type> {
            val maxType = getMaxType(level)
            return Weapon.Type.values().slice(0..maxType)
        }
    }

}